The BIG SHINY DOWNLOAD BUTTON!

laosb opened the pull request:
This download button can automatically detect your platform, and update the link for you. 

Haven’t tested on non-OS-X OS, so maybe someone should test it out.

i18n groups should translate extra keys for this button. I’m working on the zh_cn translation, and will submit it as another PR when this PR gets merged. English translation has been updated in this PR.

The `downloads.all` key is now used as `Other Downloads`, so i18n groups may need to update it to make the meaning clear.
#351

laosb committed:
The BIG SHINY DOWNLOAD BUTTON!

laosb committed:
Revert source/versions.json for the mistake.

laosb commented:
@fhemberger

snostorm commented:
Screen shot for reviewers: <img width="354" alt="iojs_screen" src="https://cloud.githubusercontent.com/assets/6534/8638499/9556c9b4-28bd-11e5-8809-0f4fea38f945.png">

I like the idea of this automatic platform button, but agree we should test it more. (I am also on OS X.)

One improvement I'd like to see is "exposing" the (existing) common list of other platform downloads versus going straight to the more rough to grok file list of downloads.

ivan commented:
Does this really make sense to do?  I would think that people frequently need to copy the Linux x64 link to their headless servers, even when browsing the website on OS X or Windows.  But I guess keeping the common platform list as mentioned in the previous comment would fix that.

laosb commented:
I guess that there are few people install io.js by `wget && tar` on their servers now? And they just need to click once more, which is not a big deal.

fhemberger commented:
@snostorm
> One improvement I'd like to see is "exposing" the (existing) common list of other platform downloads versus going straight to the more rough to grok file list of downloads.

I think we can address this with the new download template: https://github.com/nodejs/iojs.org/issues/321#issuecomment-120233986

laosb commented:
Make the link to the new download template?

fhemberger commented:
@nodejs/website Can you please comment on this change? What do you think?

robertkowalski commented on source/static/js/download.js:
```diff
@@ -0,0 +1,26 @@
+      break;
+  }
+  db.href = base + href;
+})(document, navigator);
```
- i am not a big fan of a download button based on JS - i know some people that use no-script.
- the code also misses tests, we have that problem with the js-code on nodejs.org too, it makes changes very difficult, time-consuming and error prone.
- what happens with FreeBSD people or folks that use other platforms? from following the code path in my head (again, missing tests) it looks like they get: https://iojs.org/dist/latest/undefined 
- that also means whenever there is a platform change for binaries (new platform is supported, an old platform is not supported any more), we have to remember that we have hardcoded it in JS and if we remember it, touch this piece of code, having more maintenance in the end (or bugs if we don't remember)
- for node/io i am not a user of the website for downloads, I use nvm there. but i am a user of https://www.erlang-solutions.com/downloads which also has a download button and they sniff my user agent. it always makes me mad, as I usually want the address of their .deb repo for ubuntu, but I click the button, they sniff my user agent and send me to the OSX version, where I have to switch to ubuntu manually. for me that really feels like a bad ui as i have to click 1 more time to decide against the sniffed OS. i think a list with the three major versions on https://www.erlang-solutions.com/downloads would make the page more usable and I could imagine this also applies to node/io binaries. in the case for this button submitted in the PR the user would have to find the link thats sends them to the list of supported binaries after a download for a binary started which they can't make use of.

laosb replied:
Point 1: If you don't use javascript, it is just a button directly to https://iojs.org/dist/latest/ and show `Download` as usual.
Point 2: Is it really worth writing some tests?
Point 3: A default value as '' should be ok.
Point 4: Not a very usual thing.
Point 5: Each coin has two sides. That's why having a extra link below the button. If they find that our JS mis-detected their platform, they would directly click `Other Downloads`.

If I mistaken your point, please point it out. Sorry for my bad English.

robertkowalski commented:
i have problems with a download button with logic written in JS, which is sniffing the user agent but i really like the idea of a big call-to-action-button

robertkowalski commented:
what do you think taking the nice button and linking it to: https://iojs.org/dist/v2.4.0/  - we can do that in addition to the current list

laosb commented:
@robertkowalski The button should be like that if you disabled Javascript.

laosb committed:
Default value for href.

If we can't match platforms, just link the button directly to the download list.

fhemberger commented on source/static/js/download.js:
```diff
@@ -0,0 +1,26 @@
+  var x = n.userAgent.match(/x86_64|Win64|WOW64/) ||
+    n.cpuClass === 'x64' ? 'x64' : 'x86';
+  var base = 'https://iojs.org/dist/latest/';
+  var db = d.getElementById('home-downloadbutton');
```
As I just added this on nodejs.org as well:

``` js
if (!db) { return; }
```

So this script throws no error if it is included on a page without download button.

laosb replied:
Achieved it.

laosb committed:
Avoid throwing errors without the button.

fhemberger commented:
@nodejs/website Ok everyone, what do you think: Can we merge this one?

fhemberger commented:
Okay, as I just pushed this feature for new.nodejs.org, let's have it on iojs.org as well.
